quimb/tensor/optimize.py

```diff
@@ -147,10 +153,70 @@ def parse_network_to_backend(tn, tags, constant_tags, to_constant):
         if isinstance(data, qarray):
             data = data.A
 
-        # append the raw data but mark the corresponding tensor for reinsertion
+        return data
+
+    # handle simple case of every tensor being a variable
+    if not opt_tags:
+        for t in tn_ag:
+            # append the raw data but mark the corresponding tensor
+            # for reinsertion
+            data = get_tensor_data(t)
+            variables.append(data)
+            t.add_tag(variable_tag.format(len(variables) - 1))
+        return tn_ag, variables
```

Comment:
The new refactoring is clearer but I think one still needs to parse the `constant_tags` here. Sorry the logic might get a bit messy since there are essentially two styles of specifying tags now!

1. **Opt-out** - don't specify ``tags`` and maybe specify ``constant_tags`` (and maybe ``shared_tags`` - or should this trigger opt-in style?)
2. **Opt-in** - specify ``tags`` and maybe ``shared_tags`` - probably warn if ``constant_tags`` is supplied as redundant. Might be worth having those as two functions.

This might be factored as different functions to be even clearer.

If one was being most strict (and guessing least what user wants) one could even enforce that all ``shared_tags`` appear in ``tags``, I don't know what the best course is here - thoughts welcome.

Note for the opt-in case, one can likely use this :
```python
for t in tn.select_tensors(opt_tags, which='!any'):
    # iterate over tensors which *don't* have any of the given tags
    ... make_constant(t)
```
I probably need to document all the ways one can filter tensors somewhere...

Reply:
Ahh, that is really cool that `!any` is a valid keyword for the `which`!

Reply:
The opt-in/opt-out difference is a really helpful way to think of it!

> If one was being most strict (and guessing least what user wants) one could even enforce that all shared_tags appear in tags, I don't know what the best course is here - thoughts welcome.

That makes a lot of sense, and probably also helps to clarify to the user that in the opt_in case `tags` is the set of all tags that will be optimised over and then `shared_tags` is a subset of `tags` that will be optimised together. Otherwise the overlap between the two might be confusing. It feels potentially a bit strict to raise an error when a `shared_tag` doesn't appear in `tags` though. I can imagine that being quite a big source of frustration. Perhaps a warning and a comment that it has been automatically corrected? That way the intended use is made clear without introducing too much pain

Reply:
OK that sounds good. If ``tags`` or ``shared_tags`` are given assume opt-in mode, with ``shared_tags`` supposedly a subset of ``tags``. Else default to opt-out mode.